icefoganalytics/internal-data-portal
# Visualize Tab: Limit View Not Implemented As Intended

Fixes https://github.com/icefoganalytics/internal-data-portal/issues/70

Relates to:
- https://github.com/icefoganalytics/internal-data-portal/blob/main/_Design/WireFrames%2C%20Internal%20Data%20Portal%2C%202024-01-08.pdf
- https://github.com/icefoganalytics/internal-data-portal/blob/main/_Design/User%20Stories%2C%20Internal%20Data%20Portal%2C%202024-01-08.docx
- [0004, Dataset Preview, 2024-05-21](https://docs.google.com/document/d/1wsgiX_fG8QIC07nbS4qjcuHb3IWri7xX7viqMP23xRE)

# Context

**Is your feature request related to a problem? Please describe.**
The limit data view is not meant to limit the search function rather it is to limit what a normal system user can seen in the visualization field.
![image](https://github.com/icefoganalytics/internal-data-portal/assets/89539008/d9bc21f8-4b99-4496-9960-75fb56fcb67e)

The first question is do you want to limit what data the normal user sees on the visualization tab. (I think the term “on data search” was misleading).
So if the answer is yes…they can derive what fields they want to exclude/not show in the visualization tab and how many rows it will show.

If the user has approved access these limitations are removed.

Would likely make sense to include a show no visualization as well, so that the data owner can choose to show no demo data.
I think for all the above cases this search function should be removed.

**Describe the solution you'd like**

1. Re-imagine `VisualizationControl` fields as exclusion from view, rather than exclusion from search.
    - `isDownloadableAsCsv` - stays the same
    - `hasSearchCustomizations` -> `hasPreviewCustomizations`
    - `hasFieldsExcludedFromSearch` -> `hasFieldsExcludedFromPreview` - on by default, but is a quick way to disable all field exclusions - when toggled this should probably update the DatasetField show/hide flag?
    - `hasSearchRowLimits` -> `hasPreviewRowLimit` - do we still need this?
    - `searchRowLimitMaximum` -> `previewRowLimit` - do we still need this?
    - `importSizeLimit` -> changes import size limit, though is separate from export size limit of full dataset.
`DatasetField` - `isExcludedFromSearch` -> `isExcludeFromPreview` and should be on by default for all fields when imported.
2. Add DatasetEntryPreview model to serve dataset preview that is separate from:
    i. dataset export
    ii. dataset visualization after access is granted
        Preview table is populated from DatasetEntry + DatasetField exclusions + VisualizationControl settings.

**Additional context**
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/7aa9d51b-5c8c-4cbf-8152-f76264ec3b49)

# Implementation

Add a DatasetEntryPreview model to simplify queries against a limited view of the dataset.
Vastly simplify dataset entry search, now that visualization controls don't effect search. Now it's almost simple enough to write inline due to `CROSS APPLY OPENJSON`.

## Questions/Concerns

1. It might make more sense to set up the "excluded fields" as included fields?
2. Pulling the first X records from a dataset seems like a good way to accidentally leak private data. It might make sense to have the dataset preview something that the end user must import separately? Presumably after downloading a limited view of the dataset, then removing and private information.

# Screenshots

Manager Visualize Read Preview View
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/0bc8bff5-8b94-4dd1-82f1-174b92544a98)

Manager Visualize Read Full View
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/3f33c245-b4a3-4540-8fa3-08471273ad4d)

Manager Visualize Edit Preview View
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/59f8d7a8-10a9-4fd3-a152-70982647cade)

Manager Visualize Edit Full View
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/d692484a-c9b5-475b-bf6d-553d2799985a)

User Fields Read View (with Preview access)
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/c8453131-7465-4f49-af3d-0b7406879387)

User Visualize Read View (with Preview access)
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/79d3094a-951e-46b4-8ac3-4e49697ac03c)

User Fields Read View (full access)
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/7fd11d0d-15c8-43fe-9d3c-5d4b6bda140c)

User Visualize Read View (full access)
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/d46a4cff-9cab-4956-9211-bb7228813d12)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
3. Boot the app via `dev up`
4. Log in to the app at http://localhost:8080
5. As a Data Owner add the YK Employees dataset (https://api.gov.yk.ca/directory/employees)
6. Go to the visualization tab of the dataset and you will see a note that "preview is not enabled" and a toggle to see the full dataset.
7. Toggle to see the full dataset.
8. Check that search works against the full dataset.
9. Click the "Edit" button to manage the visualization.
10. Enable preview mode, and remove at least one field from the list of excluded fields.
11. See that the table now has data in it. Search is still functional, though not very important as there are so few rows.
12. Turn preview mode off via the toggle switch to see what "full" mode look like.
13. Go to the "Access" tab and add government wide "open access".
14. In a new windows, with private browsing, log in as a new user with User role.
15. Check that you can access the dataset, and that you can see the full view of the data.
16. Using your Data Owner account, change the access grant to "Self-serve".
17. As the User account refresh the page to see the "preview" model view.
18. Go to the "Description" tab and subscribe to the dataset.
19. Go back to the visualized tab and check that you can now see the full view of the data.

## Edge cases

1. When preview is disabled, no fields or preview data should be shown to restricted user, though preview should still be available to managers via the Visualize tab.
2. When preview limit is disabled, preview should be full size of dataset.
3. When "exclude fields from preview" unchecked, all fields should be in dataset preview. Limit should still be in effect.
